metric: remove windowed histogram interface

aadityasondhi opened the pull request:
Prior to this change, the metric.Histogram maintained two seperate
histograms under the hood: cumulative and windowed. The reason for the
windowed histogram was to be able to export the most recent recordings
to TSDB. However, the windowed histogram's ticks were not synchronized
with the polling of metrics done by TSDB. This caused problems where the
TSDB polling would read an empty window and report gaps in metrics as
seen in: https://github.com/cockroachdb/cockroach/issues/98266.

This patch seperates the windowed logic from pkg/metric and makes TSDB
polling keep track of deltas using a singular cumulative histogram.

Fixes https://github.com/cockroachdb/cockroach/issues/98266
Fixes https://github.com/cockroachdb/cockroach/issues/98621

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101947)
<!-- Reviewable:end -->

aadityasondhi committed:
metric: remove windowed histogram interface

Prior to this change, the metric.Histogram maintained two seperate
histograms under the hood: cumulative and windowed. The reason for the
windowed histogram was to be able to export the most recent recordings
to TSDB. However, the windowed histogram's ticks were not synchronized
with the polling of metrics done by TSDB. This caused problems where the
TSDB polling would read an empty window and report gaps in metrics as
seen in: https://github.com/cockroachdb/cockroach/issues/98266.

This patch seperates the windowed logic from pkg/metric and makes TSDB
polling keep track of deltas using a singular cumulative histogram.

Fixes #98266
Fixes #98621

Release note: None

aadityasondhi commented:
closing in favor of https://github.com/cockroachdb/cockroach/pull/104088
